[pytree] Fix namedtuple serialization

angelayi opened the pull request:
Summary:
Previously we were serializing namedtuple treespecs incorrectly:
```python
Point = namedtuple("Point", ["x", "y"])
p = Point(1, 2)
flat, spec = pytree.tree_flatten(p)

print(flat)  # [1, 2]
print(spec)  # TreeSpec(type=namedtuple, context=Point, children=[*, *])

dumped_spec = pytree.treespec_dumps(spec)
print(dumped_spec)
"""
We only serialize the name of the class and the fields of the namedtuple:

TreeSpec {
  type='collections.namedtuple',
  context={class_name='Point', class_fields={'x', 'y'}},
  children=[Leaf, Leaf]
}
"""

reconstructed_spec = pytree.treespec_loads(dumped_spec)
print(reconstructed_spec) 
"""
When we load, we create a new namedtuple class containing the same fields as before,
but the is class is now a completely different class than the original one:

TreeSpec(type=namedtuple, context=torch.utils._pytree.Point, children=[*, *])
"""

spec == reconstructed_spec  # False
```

So, we introduce a new API called `pytree._register_namedtuple` where users can pass in the serialized name for each namedtuple class:
```python
Point = namedtuple("Point", ["x", "y"])
pytree._register_namedtuple(Point, "Point")

p = Point(1, 2)
flat, spec = pytree.tree_flatten(p)

print(flat)  # [1, 2]
print(spec)  # TreeSpec(type=namedtuple, context=Point, children=[*, *])

dumped_spec = pytree.treespec_dumps(spec)
print(dumped_spec)
"""
TreeSpec {
  type='collections.namedtuple',
  context='Point',
  children=[Leaf, Leaf]
}
"""

reconstructed_spec = pytree.treespec_loads(dumped_spec)
print(reconstructed_spec)  # TreeSpec(type=namedtuple, context=Point, children=[*, *])

spec == reconstructed_spec  # True
```

Test Plan: `python test/test_pytree.py`

Differential Revision: D55771058

zou3519 commented on torch/utils/_pytree.py:
```diff
@@ -422,18 +450,31 @@ def _namedtuple_unflatten(values: Iterable[Any], context: Context) -> NamedTuple
-    return json_namedtuple
+    if context not in SUPPORTED_SERIALIZED_TYPES:
+        raise NotImplementedError(
+            f"Serializing namedtuple class {context} in pytree is not "
+            "registered. Please register using `_register_namedtuple`",
```
nit: "Can't serialize TreeSpec of namedtuple class {context} because we didn't register a serializated_type_name. Please register using `_register_namedtuple`

zou3519 commented on torch/utils/_pytree.py:
```diff
@@ -422,18 +450,31 @@ def _namedtuple_unflatten(values: Iterable[Any], context: Context) -> NamedTuple
-    return context
+    if dumpable_context not in SERIALIZED_TYPE_TO_PYTHON_TYPE:
+        raise NotImplementedError(
+            f"Deserializing {dumpable_context} in pytree is not registered.",
```
nti: can't deserialize because no serialization name found?

angelayi committed:
[pytree] Fix namedtuple serialization (#123388)

Summary:

Previously we were serializing namedtuple treespecs incorrectly:
```python
Point = namedtuple("Point", ["x", "y"])
p = Point(1, 2)
flat, spec = pytree.tree_flatten(p)

print(flat)  # [1, 2]
print(spec)  # TreeSpec(type=namedtuple, context=Point, children=[*, *])

dumped_spec = pytree.treespec_dumps(spec)
print(dumped_spec)
"""
We only serialize the name of the class and the fields of the namedtuple:

TreeSpec {
  type='collections.namedtuple',
  context={class_name='Point', class_fields={'x', 'y'}},
  children=[Leaf, Leaf]
}
"""

reconstructed_spec = pytree.treespec_loads(dumped_spec)
print(reconstructed_spec) 
"""
When we load, we create a new namedtuple class containing the same fields as before,
but the is class is now a completely different class than the original one:

TreeSpec(type=namedtuple, context=torch.utils._pytree.Point, children=[*, *])
"""

spec == reconstructed_spec  # False
```

So, we introduce a new API called `pytree._register_namedtuple` where users can pass in the serialized name for each namedtuple class:
```python
Point = namedtuple("Point", ["x", "y"])
pytree._register_namedtuple(Point, "Point")

p = Point(1, 2)
flat, spec = pytree.tree_flatten(p)

print(flat)  # [1, 2]
print(spec)  # TreeSpec(type=namedtuple, context=Point, children=[*, *])

dumped_spec = pytree.treespec_dumps(spec)
print(dumped_spec)
"""
TreeSpec {
  type='collections.namedtuple',
  context='Point',
  children=[Leaf, Leaf]
}
"""

reconstructed_spec = pytree.treespec_loads(dumped_spec)
print(reconstructed_spec)  # TreeSpec(type=namedtuple, context=Point, children=[*, *])

spec == reconstructed_spec  # True
```

Test Plan: `python test/test_pytree.py`

Reviewed By: zou3519

Differential Revision: D55771058

XuehaiPan commented:
Related to #109460

- #109460

XuehaiPan commented on torch/utils/_pytree.py:
```diff
@@ -422,18 +450,34 @@ def _namedtuple_unflatten(values: Iterable[Any], context: Context) -> NamedTuple
+        )
+
+    serialize_node_def = SUPPORTED_SERIALIZED_TYPES[context]
+    serialized_type_name = serialize_node_def.serialized_type_name
```
Out of curious, why not automatically determine the type name via `f'{cls.__module__}.{cls.__qualname__}'` if the type is a `namedtuple`? This will not need to introduce a such new `_register_namedtuple` API.

angelayi replied:
Ah this is because this is not necessarily FC safe -- if users change the location of the defined type then it'll break the deserialization since that module will no longer exist.

angelayi commented:
@pytorchbot merge

angelayi commented:
@pytorchbot merge

desertfire commented:
This is causing failure on HUD, `torch_multimodal_clip               FAIL:     accuracy=fail_to_run, expected=pass`

angelayi commented:
Let me take a look...
